File: frontend-v2/src/api/survey.ts
import { apiGet, apiPost } from "./client";

/**
 * Yatırımcı risk profili anketi.
 *
 * SORULAR BURADA TANIMLI DEĞİL, sunucudan gelir. Tek doğruluk kaynağı
 * `data/risk_survey_config.json`; soru metnini TypeScript'e kopyalamak, ilk
 * içerik düzenlemesinde arayüzün eski soruyu sormaya devam etmesi ve skorun
 * SORULMAYAN bir soruya göre hesaplanması demek olurdu.
 *
 * Bu yüzden tipler kasıtlı olarak gevşek (`unknown`/indeks imzası): şeklin
 * ayrıntısını sabitlemek, config'i ikinci kez tanımlamak olur.
 */

/** Bir seçenek: `[kod, metin]` ya da `[kod, metin, puan]`. */
export type SurveyOption = [string, string] | [string, string, number];

export interface SurveyQuestion {
  /** Soru metni. */
  t: string;
  /** Neden sorulduğunun açıklaması (varsa) — kullanıcıya ipucu olarak gösterilir. */
  w?: string;
  o: SurveyOption[];
  /** `true` ise soru puan vermez, tavan uygular (B6). */
  cap?: boolean;
}

export interface SurveyMatrixRow {
  k: string;
  n: string;
  /** Örnek araçlar — kullanıcı hangi ürünü kastettiğimizi anlasın diye. */
  ex: string;
  w: number;
}

export interface SurveyMatrixColumn {
  k: string;
  l: string;
  o: SurveyOption[];
}

export interface SurveyProfile {
  lv: number;
  ad: string;
  lo: number;
  hi: number;
  cat: number;
  al: Record<string, number>;
}

export interface SurveyQuestions {
  meta: Record<string, unknown>;
  sorular: Record<string, SurveyQuestion>;
  urun_matrisi: { satirlar: SurveyMatrixRow[]; sutunlar: SurveyMatrixColumn[] };
  profiller: SurveyProfile[];
  arac_sirasi: string[];
  araclar: Record<string, unknown>;
}

export interface SurveyRule {
  kod: string;
  durdurucu: boolean;
  mesaj: string;
}

/**
 * Skorlama çıktısı.
 *
 * `sonuc_uretildi` `false` iken profil alanları `null` gelir — arayüz o
 * durumda profil GÖSTERMEZ, kullanıcıdan cevaplarını gözden geçirmesini ister.
 */
export interface SurveyResult {
  kapasite: number;
  tolerans: number;
  bilgi: number;
  nihai_skor: number;
  profil_seviyesi: number | null;
  profil_adi: string;
  azami_fon_risk_degeri: number | null;
  azami_urun_risk_kategorisi: number | null;
  ornek_dagilim: Record<string, number> | null;
  bilgi_nedeniyle_kisitlandi: boolean;
  likidite_tavani_uygulandi: boolean;
  sinir_bolgesinde: boolean;
  kurallar: SurveyRule[];
  sonuc_uretildi: boolean;
  /** Sonucu açıklayan kişiselleştirilmiş metin. Hiçbir zaman boş gelmez. */
  yorum: string;
}

/** Tek bir matris satırının cevabı. */
export interface SurveyMatrixAnswer {
  bilgi: string;
  siklik: string;
  hacim: string;
}

/** `{"B1": "c", "E1": {"r1": {...}}}` */
export type SurveyAnswers = Record<string, string | Record<string, SurveyMatrixAnswer>>;

export function fetchSurveyQuestions(): Promise<SurveyQuestions> {
  return apiGet<SurveyQuestions>("/api/survey/questions", { skipUnauthorizedHandler: true });
}

/**
 * Cevapları skorlar. HİÇBİR ŞEY KAYDETMEZ — kullanıcı henüz yok.
 * Kaydetme `/api/auth/register` ile ayrı bir adımda olur.
 */
export function scoreSurvey(answers: SurveyAnswers): Promise<SurveyResult> {
  return apiPost<SurveyResult>("/api/survey/score", { answers }, { skipUnauthorizedHandler: true });
}

File: backend/app/schemas/survey.py
"""Yatırımcı risk profili anketinin ekip sözleşmesi.

Sorular ve seçenekler ŞEMADA SABİT DEĞİL: `data/risk_survey_config.json`'dan
okunup olduğu gibi arayüze taşınır. Soru metnini TypeScript'e kopyalamak, tek
doğruluk kaynağı kuralını ilk düzenlemede bozardı — biri JSON'u değiştirir,
arayüz eski soruyu sormaya devam eder ve skor sorulmayan bir soruya göre
hesaplanır.

Cevap değerleri de serbest sözlük: `{"B1": "c", "E1": {"r1": {...}}}`.
Doğrulama şemada değil `survey_service` içinde yapılıyor, çünkü geçerli
seçenek kümesi yine config'ten geliyor.
"""

from typing import Any

from pydantic import BaseModel, ConfigDict


class SurveyQuestions(BaseModel):
    """Anket içeriği — arayüz formu bundan çizer.

    Config'in tamamı değil, arayüzün ihtiyacı olan bölümler: meta bilgi,
    sorular, ürün matrisi ve profil tanımları. Ağırlıklar (`o[2]`) de
    geliyor; gizlemenin bir anlamı yok (skorlama sunucuda) ve ayıklamak
    config'i iki yerde tanımlamak demek olurdu.
    """

    model_config = ConfigDict(frozen=True)

    meta: dict[str, Any]
    sorular: dict[str, Any]
    urun_matrisi: dict[str, Any]
    profiller: list[dict[str, Any]]
    arac_sirasi: list[str] = []
    araclar: dict[str, Any] = {}


class SurveyRule(BaseModel):
    """Tetiklenen tutarlılık kuralı.

    `durdurucu` ise profil ÜRETİLMEZ ve kullanıcıdan cevaplarını gözden
    geçirmesi istenir.
    """

    model_config = ConfigDict(frozen=True)

    kod: str
    durdurucu: bool
    mesaj: str


class SurveyResult(BaseModel):
    """Skorlama çıktısı.

    `profil_seviyesi` doğrudan `users.risk_survey_score` olarak yazılır
    (ikisi de 1-7). `sonuc_uretildi` `False` iken profil alanları `None`
    gelir — arayüz bu durumda profil GÖSTERMEZ.
    """

    model_config = ConfigDict(frozen=True)

    kapasite: int
    tolerans: int
    bilgi: int
    nihai_skor: int
    profil_seviyesi: int | None
    profil_adi: str
    azami_fon_risk_degeri: int | None
    azami_urun_risk_kategorisi: int | None
    ornek_dagilim: dict[str, int] | None
    bilgi_nedeniyle_kisitlandi: bool
    likidite_tavani_uygulandi: bool
    sinir_bolgesinde: bool
    kurallar: list[SurveyRule]
    sonuc_uretildi: bool
    # Sonucu kullanıcının KENDİ cevaplarına bağlayan olgular. Deterministik
    # üretilir (`survey_service._gerekceler`), LLM'e sorulmaz: hangi cevabın
    # sonucu belirlediği ölçülmüş bir bilgidir, tahmin edilemez. `yorum` bu
    # olguları cümleye döker; sağlayıcı düşerse liste yine de gelir.
    gerekceler: list[str] = []
    # Sonucu kişiselleştiren metin. LLM üretir; ulaşılamazsa profil adından
    # türetilmiş sabit bir metne düşer, ASLA boş kalmaz.
    yorum: str


class SurveyScoreRequest(BaseModel):
    """Kayıt öncesi önizleme skorlaması.

    Kimlik doğrulaması İSTEMEZ: kayıt akışının içinde, kullanıcı henüz
    yokken çağrılıyor. Hiçbir şey yazmaz, saf hesaptır.
    """

    model_config = ConfigDict(frozen=True)

    answers: dict[str, Any]
